mypyc/genops.py

```diff
@@ -1126,23 +1213,153 @@ def generate_attr_defaults(self, cdef: ClassDef) -> None:
+        # Add an attribute entry into the class dict of a non-extension class.
+        assert self.non_ext_dict is not None
+        key_unicode = self.load_static_unicode(key)
+        self.primitive_op(dict_set_item_op, [self.non_ext_dict, key_unicode, val], line)
+
+    def add_non_ext_class_attr(self, lvalue: NameExpr,
+                               stmt: AssignmentStmt, class_ir: ClassIR) -> None:
+        # We populate __annotations__ because dataclasses uses it to determine
+        # which attributes to compute on.
+        # TODO: Maybe generate more precise types for annotations
+        assert self.non_ext_anns is not None
+
+        key = self.load_static_unicode(lvalue.name)
+        typ = self.primitive_op(type_object_op, [], stmt.line)
+        self.primitive_op(dict_set_item_op, [self.non_ext_anns, key, typ], stmt.line)
+
+        # Only add the attribute to the __dict__ if the assignment is of the form:
+        # x : type = value (don't add attributes of the form 'x : type' to the __dict__).
+        if not isinstance(stmt.rvalue, TempNode):
+            self.add_to_non_ext_dict(lvalue.name, self.accept(stmt.rvalue), stmt.line)
+
+    def setup_non_ext_dict(self, cdef: ClassDef) -> Value:
+        # Initialize the class dictionary for a non-extension class. This class dictionary
+        # is passed to the metaclass constructor.
+        assert self.non_ext_bases is not None
+
+        # Check if the metaclass defines a __prepare__ method, and if so, call it.
+        metaclass = self.primitive_op(type_object_op, [], cdef.line)
+        metaclass = self.primitive_op(py_calc_meta_op, [metaclass, self.non_ext_bases], cdef.line)
+        has_prepare = self.primitive_op(py_hasattr_op,
+                                        [metaclass,
+                                        self.load_static_unicode('__prepare__')], cdef.line)
+
+        non_ext_dict = self.alloc_temp(dict_rprimitive)
+
+        true_block, false_block, exit_block, = BasicBlock(), BasicBlock(), BasicBlock()
+        self.add_bool_branch(has_prepare, true_block, false_block)
+
+        self.activate_block(true_block)
+        cls_name = self.load_static_unicode(cdef.name)
+        prepare_meth = self.py_get_attr(metaclass, '__prepare__', cdef.line)
+        prepare_dict = self.py_call(prepare_meth, [cls_name, self.non_ext_bases], cdef.line)
+        self.assign(non_ext_dict, prepare_dict, cdef.line)
+        self.goto(exit_block)
+
+        self.activate_block(false_block)
+        self.assign(non_ext_dict, self.primitive_op(new_dict_op, [], cdef.line), cdef.line)
+        self.goto(exit_block)
+        self.activate_block(exit_block)
+
+        return non_ext_dict
+
     def visit_class_def(self, cdef: ClassDef) -> None:
-        self.allocate_class(cdef)
+        ir = self.mapper.type_to_ir[cdef.info]
+        # Currently, we only create non-extension classes for classes that are
+        # decorated. Classes decorated with @trait do not apply here, and are
```

Comment:
Should we add "or have a custom metaclass other than \<list of supported metaclasses here\>"?